### PR TITLE
Sync releases table with tags on push and for mirrors

### LINKED ISSUE #2154: Releases didn't display
see https://try.gitea.io/lunny/core/releases for examples. This repo is migrated from https://github.com/go-xorm/core. The counter is 8, but nothing listed.

### PR DESCRIPTION
**Pull request by lafriks**
Fixes #2154

~~Not fully tested but ready for review~~

### CODE REVIEW

File: models/release.go
Comment (ethantkoenig) [inline models/release.go]: Could we use a `map[string]struct{}` instead? It will be more efficient to check if tags have been checked below with a map.

File: routers/api/v1/repo/release.go
Comment (ethantkoenig) [inline routers/api/v1/repo/release.go]: Typo?

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

File: models/migrations/v39.go
Comment (lunny) [inline models/migrations/v39.go]: Why update all are `false` ?

File: models/migrations/v39.go
Comment (lunny) [inline models/migrations/v39.go]: offset += len(repos)

File: routers/api/v1/repo/release.go
Comment (lafriks) [reply routers/api/v1/repo/release.go]: If it is typo than it was also previously as I kept previous behaviour, it was returning status 409 when release already existed

File: models/migrations/v39.go
Comment (lafriks) [reply models/migrations/v39.go]: default, probably not needed

File: models/release.go
Comment (lunny) [inline models/release.go:409]: update after delete?

Comment (lafriks) [reply models/release.go:409]: forgot to remove update, good catch, thx

Comment (lafriks) [reply models/release.go:409]: actually it is just in diff looks like this, if you look at the full file code there is no update before delete, update is from other function :)

File: routers/api/v1/repo/release.go
Comment (lafriks) [reply routers/api/v1/repo/release.go]: I think it is not a typo: https://github.com/go-gitea/gitea/blob/93a1de48428c2f89502cf6bba80f3976c6f5839f/routers/api/v1/repo/release.go#L97

**Comment by bkcsoft** [comment]
So, why do we need to do this? Tags and Releases are not the same thing, they should not be treated as the same thing... If we need this list it should be fetched from the GitRepo directly 🤔

**Comment by lafriks** [comment]
@bkcsoft this is only way to efficiently implement release page paging and to have release page match github behavior. Releases are actually just extenended tags so can say they are quite the same

**Comment by lunny** [comment]
@lafriks I think maybe @bkcsoft is right. Since user will input username and password, we can retrieves release information via github API and inserted to gitea.( maybe an checkbox option: sync releases ). And this sync function also could be a button on `tags` UI to publish all the tags.

**Comment by lafriks** [comment]
@bkcsoft @lunny it's not about just migration from github it's about displaying tags that are not linked to releases in release page like github is doing:

image

And it does not matter if repository is migrated or tags are created from git client. Not everyone wants or need to create releases for each tag but it is useful when creating tag to see kind of release list and only if needed create release with additional info and attachments. This is especially important for mirrored repositories and they can also be mirrored from github/gitlab etc

**Comment by lafriks** [comment]
And also this was the functionality gitea had before release page was rewritten to show only data from database (I'm not against it)

**Comment by lunny** [comment]
@lafriks OK. I see.

**Comment by lunny** [comment]
LGTM

**Comment by bkcsoft** [comment]
> we can retrieves release information via github API

@lunny I thought that any non-repo migration/sync would be left to third-party tools?

**Comment by lafriks** [comment]
@bkcsoft but this PR is not related to it in any way but otherwise I agree 3rd-party sync/migration should be external tool

File: models/release.go
Comment (ethantkoenig) [inline models/release.go]: Should this be `"is_tag"`?

File: routers/api/v1/repo/release.go
Comment (ethantkoenig) [inline routers/api/v1/repo/release.go]: I'm confused here. If we are creating a release, why do we expect there to already be a release with the name `form.TagName`?

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

**Comment by bkcsoft** [comment]
@lafriks Mainly just thinking out loud 🙂

**Comment by lafriks** [comment]
@ethantkoenig fixed and also refactored code to minimize chance of possible race conditions

File: routers/api/v1/repo/release.go
Comment (lafriks) [reply routers/api/v1/repo/release.go]: because release table does now contains also tags that are not full releases with attachments etc and when creating release it is needed to update release table flag (is_tag to false) and adding additional release information to that row

**Comment by lunny** [comment]
@lafriks and this one.

**Comment by lafriks** [comment]
@lunny I think I found bug but have to still reproduce that, do not merge yet. Would be great if someone could test it on his dev environment

File: models/release.go
Comment (ethantkoenig) [inline models/release.go]: nit: `for page := 1; ; page++ {`

**Review by ethantkoenig — APPROVED** [review]

**Comment by lafriks** [comment]
@ethantkoenig fixed your suggestion
@lunny I think I fixed the bug but would appreciate if someone could test it more. At least for me it seems to work correctly now

**Comment by daviian** [comment]
@lafriks Found two issues.

1. git treats tags case sensitive, therefore using `toLowerCase` could drop a tag
2. if a tag is removed and added on a new commit sync doesn't update tag reference

But maybe there are more.

**Comment by daviian** [comment]
@lafriks Guess I found another one.

Creating a release with an existing tag doesn't work anymore because an existing tag is already present as release and database is complaining about unique constraint.

**Comment by daviian** [comment]
@lafriks Ok, I have finished my examination. Last commit definitely fixes bugs regarding release creation.
So above mentioned issues are still a problem.

As `toLowerCase` isn't an issue created with this PR it should be addressed separately.
But IMO tag reference update should be fixed before merging.

**Comment by ethantkoenig** [comment]
~@lafriks I believe I found a bug.~ Never mind, was running an old version.

Steps to reproduce:
1. Create a mirror repository of a Github repo with releases (for example, I used https://github.com/ethantkoenig/run_lambda)
2. Make the repository a regular, non-mirror repository
3. Clone the repo, and try deleting a tag (e.g. `git push origin :v1.0.4`)
4. The release count remains unchanged, and the deleted tag still shows up in the list of release. When you try to view the deleted tag, you get a 404.

**Review by ethantkoenig — CHANGES_REQUESTED** [review]
Never mind

**Review by ethantkoenig — APPROVED** [review]

**Comment by ethantkoenig** [comment]
LGTM

**Comment by ethantkoenig** [comment]
Make L-G-T-M work

**Comment by lunny** [comment]
I made a mistake. @lafriks

**Pull request merged**
